### PR TITLE
Send events one call after renaming file

### PR DESCRIPTION
**Pull request by martijnarts**
*Will not actually be merging this, just looking for code review before I PR it for Segment.*

This solves a race condition where the client can write to the file after the sender has already renamed it and read its contents. This would mean that events can (and do) disappear with the `unlink` call. In our production environment we lost over 30 events in an hour.

Graphical explanation:
```
  client │ open   write(1)   write(2)
         │ ↓      ↓          ↓
   $file ├──────────────────────────────...───...
         │   ↑        ↑                     ↑
send.php │   rename   file_get_contents()   unlink()
```

`write(1)` is fine, since the file descriptor will persist after the rename. `write(2)` is will be lost. The solution here is decreasing the possibility of this happening by assuming that clients don't live longer than the `save.php` cron interval (a minute, in our case):

1. Rename the current file.
2. Read the directory for files _other than_ the just-renamed file, parse those events, remove them.

This means that the current file is left, renamed, for the duration of the cron interval. So basically a grace period of a minute. This is the most effective way I could think of of reducing the issue without changing the client.

### CODE REVIEW

File: send.php [hunk 1]
```diff
@@ -29,64 +29,73 @@
...
-  print("file: $old does not exist");
-  exit(0);
-}
+$newFileName = 'analytics-' . rand() . '.log';
+$newFile = $dir . '/' . $newFileName;
 
-if (!rename($old, $file)) {
-  print("error renaming from $old to $file\n");
-  exit(1);
+if(file_exists($old)) {
```
Comment (artemgordinskiy) [inline send.php:36]: So if old file does not exist, we won't print anything? Could be nice to still do it for debugging.

File: send.php [hunk 2]
```diff
@@ -29,64 +29,73 @@
...
-  $payload = json_decode($line, true);
-  $dt = new DateTime($payload["timestamp"]);
-  $ts = floatval($dt->getTimestamp() . "." . $dt->format("u"));
-  $payload["timestamp"] = $ts;
-  $type = $payload["type"];
-  $ret = call_user_func_array(array("Segment", $type), array($payload));
-  if ($ret) $successful++;
-  $total++;
-  if ($total % 100 === 0) Segment::flush();
+foreach(scandir($dir) as $file) {
```
Comment (artemgordinskiy) [inline send.php]: Could be a good idea to add `SCANDIR_SORT_NONE` as the second argument to avoid wasting time on sorting if there are a lot of files.
Also, `scandir()` is going to generate a warning if directory does not exist (which may be true if file does not exist).

File: send.php [hunk 3]
```diff
@@ -29,64 +29,73 @@
...
+    */
+
+    $contents = file_get_contents($file);
+    $lines = explode("\n", $contents);
+
+    /**
+    * Initialize the client.
+    */
+
+    Segment::init($args["secret"], array(
```
Comment (artemgordinskiy) [inline send.php]: Might want to move Segment initialisation out of the loop since we might end up doing it multiple times.

File: send.php [hunk 4]
```diff
@@ -29,64 +29,73 @@
...
-  $dt = new DateTime($payload["timestamp"]);
-  $ts = floatval($dt->getTimestamp() . "." . $dt->format("u"));
-  $payload["timestamp"] = $ts;
-  $type = $payload["type"];
-  $ret = call_user_func_array(array("Segment", $type), array($payload));
-  if ($ret) $successful++;
-  $total++;
-  if ($total % 100 === 0) Segment::flush();
+foreach(scandir($dir) as $file) {
+    if (substr($file, 0, strlen('analytics-')) !== 'analytics-') {
```
Comment (artemgordinskiy) [inline send.php]: 1. This feels kinda fragile.
A. What if there's a lot unrelated files starting with `analytics-`?
B. How do we know that those files start with `analytics-` in the first place?
C. If a cron run takes more than a minute and another one starts up in parallel, you could probably end up with duplicate events.
2. You could also write this as `strpos($file, 'analytics-') !== 0`

**Review by artemgordinskiy — CHANGES_REQUESTED** [review]
A couple of things

File: send.php (hunk 1 above)
Comment (martijnarts) [reply send.php:36]: Don't think it's a "broken" scenario in any way: there's a very legitimate scenario that the new file simply has not been created yet. Generally the printing seems to be output status details only.
Could print it, but I'm not sure if it adds anything necessarily.

File: send.php (hunk 2 above)
Comment (martijnarts) [reply send.php]: Just a warning? It should break...

File: send.php (hunk 4 above)
Comment (martijnarts) [reply send.php]: 1B. because we rename them to that :)
1A. that's fair I suppose, but I don't really see a solution to this?
1C. Very good catch! I guess we can have a lockfile (`analytics.lock`?) in the same directory to prevent two crons running at the same time.
2. Yes.

Comment (artemgordinskiy) [reply send.php]: 1B. My bad, didn't really get how it worked.
1A. Maybe we could try using a more unique name? Like a hash of the current directory path or something like that.
1C. Yeah, or we could also use [flock](https://linux.die.net/man/1/flock)

File: send.php (hunk 2 above)
Comment (artemgordinskiy) [reply send.php]: Yeah, `scandir()` will also return `false` which will also trigger a warning from `foreach` I think.

**Review by artemgordinskiy — APPROVED** [review]
:shipit:
